include/dlaf/eigensolver/tridiag_solver/mc.h

```diff
@@ -0,0 +1,241 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2022, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+#pragma once
+
+#include <algorithm>
+
+#include <pika/future.hpp>
+#include <pika/unwrap.hpp>
+
+#include "dlaf/common/callable_object.h"
+#include "dlaf/eigensolver/tridiag_solver/api.h"
+#include "dlaf/eigensolver/tridiag_solver/merge.h"
+#include "dlaf/lapack/tile.h"
+#include "dlaf/matrix/copy_tile.h"
+#include "dlaf/sender/make_sender_algorithm_overloads.h"
+#include "dlaf/sender/policy.h"
+#include "dlaf/sender/transform.h"
+#include "dlaf/types.h"
+
+namespace dlaf::eigensolver::internal {
+
+template <class T>
+struct TridiagSolver<Backend::MC, Device::CPU, T> {
+  static void call(Matrix<T, Device::CPU>& mat_trd, Matrix<T, Device::CPU>& d,
+                   Matrix<T, Device::CPU>& mat_ev);
+};
+
+/// Splits [i_begin, i_end] in the middle and waits for all splits on [i_begin, i_middle] and [i_middle +
+/// 1, i_end] before saving the triad <i_begin, i_middle, i_end> into `indices`.
+///
+/// The recursive calls span a binary tree which is traversed in depth first left-right-root order. That
+/// is also the order of triads in `indices`.
+///
+/// Note: the intervals are all closed!
```

Comment:
I suppose this also means you will never need empty intervals? If yes, would some assertions about this be in place?

Reply:
By empty intervals, do you mean e.g. `[0, 0]` or `[]`? The former is a valid interval that contains only `0`, the latter is not something I need and will perhaps benefit from an assertion.

I have generally been lax on assertions with internal parts of the code.

Reply:
I meant really empty (`[]`) just because closed intervals make representing those impossible/difficult (unless you allow something like `[i, i-1]`). But if you indeed don't need those then a few assertions might be useful.